src/mpid/ch4/netmod/ucx/ucx_recv.h

```diff
@@ -8,30 +8,16 @@
 
 #include "ucx_impl.h"
 
-/* when the recv is immediately completed, we need get a request to fill the status,
- * but we need to get request from the correct pool or race condition arises. And we
- * don't know the correct vni!
- * The new ucx API should help: https://github.com/openucx/ucx/pull/5060
- * For now, we malloc a temporary request.
- * FIXME: update with new ucx api
- */
 MPL_STATIC_INLINE_PREFIX void MPIDI_UCX_recv_cmpl_cb(void *request, ucs_status_t status,
-                                                     ucp_tag_recv_info_t * info)
+                                                     const ucp_tag_recv_info_t * info,
+                                                     void *user_data)
 {
```

Comment:
Just to understand, with the new API, the callback is always called, right?

Reply:
In the recv case, yes, because we force it. See the `UCP_OP_ATTR_FLAG_NO_IMM_CMPL` attribute added the operation parameters. The send side can still complete immediately, however. There may be some improvement we can do on that side by first trying immediate completion, and falling back to allocating a request and forcing a callback. I'm not sure it makes as much difference as it does on recv, though.